test/common/router/router_test.cc

```diff
@@ -381,6 +381,42 @@ TEST_F(RouterTest, UpstreamPerTryTimeout) {
   EXPECT_EQ(1UL, cm_.conn_pool_.host_->stats().rq_timeout_.value());
 }
 
+TEST_F(RouterTest, PerTryTimeoutWithNoUpstreamHost) {
+  NiceMock<Http::MockStreamEncoder> encoder;
+  Http::StreamDecoder* response_decoder = nullptr;
+  EXPECT_CALL(cm_.conn_pool_, newStream(_, _))
+      .WillOnce(Invoke([&](Http::StreamDecoder& decoder, Http::ConnectionPool::Callbacks& callbacks)
+                           -> Http::ConnectionPool::Cancellable* {
+        response_decoder = &decoder;
+        // simulate connect timeout, do not call callbacks.onPoolReady(...)
+        UNREFERENCED_PARAMETER(callbacks);
+        return nullptr;
```

Comment:
nit: Can you return a real Cancellable here to make it more realistic. You should be able to expect the cancel() call. (See other tests).